docs/proposals/20230608-nri-mode-resource-management.md

```diff
@@ -0,0 +1,175 @@
+- Transform the runtime hook plugins' response into OCI spec format 
+- Return OCI spec format response to container runtime(e.g. containerd, crio).
+
+![img.png](../images/nri-proposal.png)
+
+### User Stories
+
+#### Story 1
+As a cluster administrator, I want to apply QoS policy before pod's status become running.
+
+#### Story 2
+As a cluster administrator, I want to deploy koordinator cluster without restart.
+
+#### Story 3
+As a cluster administrator, I want to adjust resources' policies at runtime.
+
+#### Story 4
+As a GPU user, I want to inject environment before pod running.
+
+### Requirements (Optional)
+
+- Need to upgrade containerd to >= 1.7.0, crio to v1.25.0
+
+#### Functional Requirements
+
+NRI mode should support all existing functionalities supported by standalone and Proxy mode.
+
+#### Non-Functional Requirements
+
+Non-functional requirements are user expectations of the solution. Include
+considerations for performance, reliability and security.
+
+### Implementation Details/Notes/Constraints
+1. koordlet [NRI plugin](https://github.com/containerd/nri/blob/main/plugins/template/plugin.go)
+```go
+type nriServer struct {
+	stub    stub.Stub
+	mask    stub.EventMask
+	options Options // server options
+}
+
+// Enable 3 hooks (RunPodSandbox, CreateContainer, UpdateContainer) in NRI 
+func (p *nriServer) Configure(config, runtime, version string) (stub.EventMask, error) {
+}
+
+// Sync all pods/containers information before koordlet nri plugin run
+func (p *nriServer) Synchronize(pods []*api.PodSandbox, containers []*api.Container) ([]*api.ContainerUpdate, error) {
+}
+
+func (p *nriServer) RunPodSandbox(pod *api.PodSandbox) error {
+	podCtx.FromNri(pod)
+	RunHooks(...)
+	podCtx.NriDone()
+}
+
+func (p *nriServer) CreateContainer(pod *api.PodSandbox, container *api.Container) (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+    containerCtx.FromNri(pod, container)
+	RunHooks(...)
+	containCtx.NriDone()
+}
+
+func (p *nriServer) UpdateContainer(pod *api.PodSandbox, container *api.Container) ([]*api.ContainerUpdate, error) {
+    containerCtx.FromNri(pod, container)
+	RunHooks(...)
+	containCtx.NriDone()
+}
+```
+2. koordlet enhancement for NRI
+- PodContext
+```go
+// fill PodContext from OCI spec
+func (p *PodContext) FromNri(pod *api.PodSandbox) {
+}
+
+// apply QoS resource policies for pod
+func (p *PodContext) NriDone() {
+}
+```
+- ContainerContext
+```go
+// fill ContainerContext from OCI spec
+func (c *ContainerContext) FromNri(pod *api.PodSandbox, container *api.Container) {
+}
+
+// apply QoS resource policies for container
+func (c *ContainerContext) NriDone() (*api.ContainerAdjustment, []*api.ContainerUpdate, error) {
+}
+```
+
+### Risks and Mitigations
+
+## Upgrade Strategy
+
+- Need to upgrade containerd to 1.7.0+ or CRIO to 1.25.0+
+- Need to enable NRI
+
+### Test Plan [optional]
+
```

Comment:
As we discussed in the biweekly meeting, please add an *Alternative* section for comparisons with the runtime-proxy mode.

Reply:
Yes, added. Compare runtime-proxy with NRI briefly now.

Reply:
> As we discussed in the biweekly meeting, please add an _Alternative_ section for comparisons with the runtime-proxy mode.

Also added the comparisons with standalone mode.